File: go-probe/registry.go
// Реестр задач по расписанию + cron-планирование — аналог C# Worker и CronExecuter.
//
// Сервер присылает только изменения (MergeJobs): новые задачи добавляются,
// существующие обновляются, помеченные Delete — удаляются. Разовые задачи
// (Repeater) выполняются сразу и в реестре не хранятся. Реестр переживает
// перезапуск (TaskInfo.json). Каждая задача планируется таймером на момент
// следующего срабатывания cron-выражения; по срабатыванию задача уходит
// в очередь диспетчера, и сразу планируется следующий запуск.
package main

import (
	"encoding/json"
	"maps"
	"os"
	"slices"
	"sync"
	"time"

	"github.com/robfig/cron/v3"
)

// tasksFileName — файл с сохранённым реестром задач по расписанию.
const tasksFileName = "TaskInfo.json"

// cronStandard разбирает классические 5-польные cron-выражения.
var cronStandard = cron.NewParser(
	cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)

// cronWithSeconds разбирает 6-польные выражения с секундами.
var cronWithSeconds = cron.NewParser(
	cron.Second | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)

// scheduledTask — одна задача по расписанию с её таймером.
type scheduledTask struct {
	task  *TaskInfo
	timer *time.Timer
}

// Enqueuer — приёмник задач на выполнение (реализуется диспетчером;
// в тестах подменяется заглушкой).
type Enqueuer interface {
	Enqueue(task *TaskInfo)
}

// TaskRegistry — реестр задач по расписанию.
type TaskRegistry struct {
	mu         sync.Mutex
	tasks      map[string]*scheduledTask
	dispatcher Enqueuer
	registry   *RunRegistry
	cancels    *RunCancelRegistry // чтобы оборвать выполнение удалённой задачи
}

// NewTaskRegistry создаёт пустой реестр.
func NewTaskRegistry(dispatcher Enqueuer, registry *RunRegistry, cancels *RunCancelRegistry) *TaskRegistry {
	return &TaskRegistry{tasks: map[string]*scheduledTask{}, dispatcher: dispatcher, registry: registry, cancels: cancels}
}

// Load восстанавливает реестр из TaskInfo.json после перезапуска.
func (r *TaskRegistry) Load() {
	data, err := os.ReadFile(tasksFileName)
	if err != nil {
		return // файла нет — чистая проба, задачи дошлёт сверка сервера
	}
	var saved []TaskInfo
	if err := json.Unmarshal(data, &saved); err != nil {
		logRegistry.Error("Не удалось загрузить реестр задач", "файл", tasksFileName, "ошибка", err)
		return
	}
	r.MergeJobs(saved)
	logRegistry.Info("Реестр задач восстановлен с диска", "файл", tasksFileName, "задач", len(saved))
}

// MergeJobs применяет инкрементальные изменения задач (добавить/обновить/удалить).
func (r *TaskRegistry) MergeJobs(jobs []TaskInfo) {
	r.mu.Lock()
	defer r.mu.Unlock()

	changed := false
	for i := range jobs {
		changed = r.mergeOne(&jobs[i]) || changed
	}
	if changed {
		r.persist()
	}
}

// mergeOne применяет одну задачу; вызывается под mu. Возвращает «реестр изменился».
func (r *TaskRegistry) mergeOne(item *TaskInfo) bool {
	// Пришло удаление — прежде всего обрываем то, что уже выполняется.
	// Снять расписание мало: зонд вроде «twping -c 300» живёт минутами и иначе
	// домерял бы по несуществующей задаче. Делаем это для любой задачи, даже
	// если её нет в реестре: разовые (Repeater) там не хранятся вовсе.
	stopped := 0
	if item.Delete {
		stopped = r.cancels.CancelTask(item.Id)
	}

	// Разовые задачи выполняем немедленно и в реестре не храним.
	if item.Type == TypeRepeater {
		if !item.Delete {
			r.dispatcher.Enqueue(item)
		} else if stopped > 0 {
			logRegistry.Info("Разовая задача удалена — выполнение прервано",
				"задача", item.Id, "название", item.Title, "оборвано_запусков", stopped)
		}
		return false
	}

	// Задача по расписанию, помеченная на удаление.
	if item.Delete {
		entry, ok := r.tasks[item.Id]
		if !ok {
			// В реестре её уже нет, но выполнение мы прервали — это главное.
			return false
		}
		if entry.timer != nil {
			entry.timer.Stop()
		}
		delete(r.tasks, item.Id)
		r.registry.Remove(item.Id)
		logRegistry.Info("Задача удалена", "задача", item.Id, "название", item.Title,
			"оборвано_запусков", stopped)
		return true
	}

	// Обновление существующей или добавление новой.
	if entry, ok := r.tasks[item.Id]; ok {
		if entry.timer != nil {
			entry.timer.Stop()
		}
		entry.task = item
		r.scheduleNext(entry)
		logRegistry.Debug("Задача обновлена", "задача", item.Id, "название", item.Title)
		return true
	}

	entry := &scheduledTask{task: item}
	r.tasks[item.Id] = entry
	r.scheduleNext(entry)
	logRegistry.Debug("Задача добавлена", "задача", item.Id, "название", item.Title,
		"режим", item.Mode, "узел", item.EndNode, "расписание", item.CronExpression)
	return true
}

// scheduleNext планирует следующий запуск задачи; вызывается под mu.
func (r *TaskRegistry) scheduleNext(entry *scheduledTask) {
	task := entry.task

	parser := cronStandard
	if task.CronWithSeconds {
		parser = cronWithSeconds
	}
	schedule, err := parser.Parse(task.CronExpression)
	if err != nil {
		logRegistry.Error("Некорректное cron-выражение — задача не будет запускаться",
			"задача", task.Id, "название", task.Title, "выражение", task.CronExpression, "ошибка", err)
		r.registry.SetNextRun(task.Id, task.Title, string(task.Mode), nil)
		return
	}

	next := schedule.Next(time.Now())
	if !task.End.IsZero() && next.After(task.End.Time) {
		logRegistry.Info("Задача завершена по дате окончания",
			"задача", task.Id, "название", task.Title, "окончание", task.End.Format("02.01.2006 15:04"))
		r.registry.SetNextRun(task.Id, task.Title, string(task.Mode), nil)
		return
	}

	// Фиксируем план в реестре статусов — оператор видит, когда следующий запуск.
	r.registry.SetNextRun(task.Id, task.Title, string(task.Mode), &next)

	entry.timer = time.AfterFunc(time.Until(next), func() {
		// Только ставим в очередь — выполнение возьмёт диспетчер; следующее
		// срабатывание планируем сразу, не дожидаясь завершения зонда.
		r.dispatcher.Enqueue(task)

		r.mu.Lock()
		defer r.mu.Unlock()
		if current, ok := r.tasks[task.Id]; ok && current == entry {
			r.scheduleNext(entry)
		}
	})
}

// KnownTaskIds возвращает идентификаторы задач по расписанию (для сверки сервером).
func (r *TaskRegistry) KnownTaskIds() []string {
	r.mu.Lock()
	defer r.mu.Unlock()
	// Итератор ключей + сборка в срез — Go 1.23. Пустой результат обязан быть
	// «[]», а не «null»: сервер разворачивает ответ в HashSet без проверки на null.
	ids := slices.Collect(maps.Keys(r.tasks))
	if ids == nil {
		ids = []string{}
	}
	return ids
}

// AllTasks возвращает полные определения задач по расписанию, которые держит проба.
// Сервер забирает их для восстановления своей БД после потери данных.
func (r *TaskRegistry) AllTasks() []TaskInfo {
	r.mu.Lock()
	defer r.mu.Unlock()
	all := make([]TaskInfo, 0, len(r.tasks))
	for _, entry := range r.tasks {
		all = append(all, *entry.task)
	}
	return all
}

// ClearAll останавливает и удаляет ВСЕ задачи по расписанию вместе с файлом реестра.
// Используется сторожем связи: сервер молчит дольше таймаута — проба считает себя
// удалённой. Возвращает число остановленных задач.
func (r *TaskRegistry) ClearAll() int {
	r.mu.Lock()
	defer r.mu.Unlock()

	count := len(r.tasks)
	for id, entry := range r.tasks {
		if entry.timer != nil {
			entry.timer.Stop()
		}
		delete(r.tasks, id)
		r.registry.Remove(id)
	}
	_ = os.Remove(tasksFileName)
	// Заодно обрываем всё, что выполняется прямо сейчас.
	if stopped := r.cancels.CancelAll(); stopped > 0 {
		logRegistry.Info("Оборваны выполняющиеся запуски", "запусков", stopped)
	}
	return count
}

// persist сохраняет реестр на диск; вызывается под mu.
func (r *TaskRegistry) persist() {
	all := make([]TaskInfo, 0, len(r.tasks))
	for _, entry := range r.tasks {
		all = append(all, *entry.task)
	}
	data, err := json.Marshal(all)
	if err != nil {
		logRegistry.Error("Не удалось сериализовать реестр задач", "ошибка", err)
		return
	}
	if err := os.WriteFile(tasksFileName, data, 0o644); err != nil {
		logRegistry.Error("Не удалось сохранить реестр задач", "файл", tasksFileName, "ошибка", err)
	}
}

File: rust-probe/src/registry.rs
//! Реестр задач по расписанию + cron-планирование — аналог Go TaskRegistry,
//! C# Worker и CronExecuter.
//!
//! Сервер присылает только изменения (merge_jobs): новые задачи добавляются,
//! существующие обновляются, помеченные Delete — удаляются. Разовые задачи
//! (Repeater) выполняются сразу и в реестре не хранятся. Реестр переживает
//! перезапуск (TaskInfo.json). Единый поток-планировщик будит задачи по времени
//! ближайшего срабатывания cron-выражения.

use crate::contracts::{CsTime, TaskInfo, TaskType};
use crate::dispatcher::Dispatcher;
use crate::logging::{f, Logger};
use crate::runregistry::RunRegistry;
use chrono::{DateTime, Local};
use std::collections::HashMap;
use std::str::FromStr;
use std::sync::{Arc, Condvar, Mutex};
use std::time::Duration;

const TASKS_FILE: &str = "TaskInfo.json";
static LOG: Logger = Logger("registry");

/// Одна задача по расписанию с моментом следующего запуска.
struct Scheduled {
    task: Arc<TaskInfo>,
    next: Option<DateTime<Local>>,
}

/// Реестр задач по расписанию.
pub struct TaskRegistry {
    tasks: Mutex<HashMap<String, Scheduled>>,
    wake: Condvar,
    dispatcher: Mutex<Option<Arc<Dispatcher>>>,
    registry: Arc<RunRegistry>,
}

impl TaskRegistry {
    pub fn new(registry: Arc<RunRegistry>) -> Arc<Self> {
        Arc::new(TaskRegistry {
            tasks: Mutex::new(HashMap::new()),
            wake: Condvar::new(),
            dispatcher: Mutex::new(None),
            registry,
        })
    }

    /// Привязывает диспетчер (создаётся позже реестра) и запускает поток-планировщик.
    pub fn start(self: &Arc<Self>, dispatcher: Arc<Dispatcher>) {
        *self.dispatcher.lock().unwrap() = Some(dispatcher);
        let me = Arc::clone(self);
        std::thread::spawn(move || me.scheduler_loop());
    }

    /// Ставит задачу в очередь диспетчера.
    fn enqueue(&self, task: Arc<TaskInfo>) {
        if let Some(d) = self.dispatcher.lock().unwrap().as_ref() {
            d.enqueue(task);
        }
    }

    /// Восстанавливает реестр задач с диска.
    pub fn load(self: &Arc<Self>) {
        let Ok(data) = std::fs::read(TASKS_FILE) else {
            return; // файла нет — чистый старт
        };
        let saved: Vec<TaskInfo> = match serde_json::from_slice(&data) {
            Ok(v) => v,
            Err(e) => {
                LOG.error("Не удалось загрузить реестр задач", &[f("файл", TASKS_FILE), f("ошибка", e)]);
                return;
            }
        };
        let count = saved.len();
        self.merge_jobs(saved);
        if count > 0 {
            LOG.info("Реестр задач восстановлен с диска", &[f("файл", TASKS_FILE), f("задач", count)]);
        }
    }

    /// Сливает присланные изменения задач в реестр.
    pub fn merge_jobs(self: &Arc<Self>, jobs: Vec<TaskInfo>) {
        let mut changed = false;
        {
            let mut tasks = self.tasks.lock().unwrap();
            for job in jobs {
                changed |= self.merge_one(&mut tasks, job);
            }
            if changed {
                persist(&tasks);
            }
        }
        if changed {
            self.wake.notify_all(); // планировщик пересчитает ближайший срок
        }
    }

    /// Применяет одну задачу; вызывается под замком. Возвращает «реестр изменился».
    fn merge_one(&self, tasks: &mut HashMap<String, Scheduled>, item: TaskInfo) -> bool {
        // Разовые задачи выполняем немедленно и в реестре не храним.
        if item.task_type == TaskType::Repeater {
            if !item.delete {
                self.enqueue(Arc::new(item));
            }
            return false;
        }

        // Задача по расписанию, помеченная на удаление.
        if item.delete {
            if tasks.remove(&item.id).is_some() {
                self.registry.remove(&item.id);
                LOG.info("Задача удалена", &[f("задача", item.id.clone()), f("название", item.title.clone())]);
                return true;
            }
            return false;
        }

        let existed = tasks.contains_key(&item.id);
        let task = Arc::new(item);
        let next = self.compute_next(&task);
        tasks.insert(task.id.clone(), Scheduled { task: Arc::clone(&task), next });

        if existed {
            LOG.debug("Задача обновлена", &[f("задача", task.id.clone()), f("название", task.title.clone())]);
        } else {
            LOG.debug(
                "Задача добавлена",
                &[
                    f("задача", task.id.clone()),
                    f("название", task.title.clone()),
                    f("режим", task.mode.as_str()),
                    f("узел", task.end_node.clone()),
                    f("расписание", task.cron_expression.clone()),
                ],
            );
        }
        true
    }

    /// Вычисляет момент следующего запуска по cron-выражению задачи.
    fn compute_next(&self, task: &TaskInfo) -> Option<DateTime<Local>> {
        // Крейт cron ждёт 6–7 полей (с секундами); классическое 5-польное выражение
        // дополняем секундами «0», как это делает парсер Go-пробы.
        let expr = if task.cron_with_seconds {
            task.cron_expression.clone()
        } else {
            format!("0 {}", task.cron_expression.trim())
        };

        let schedule = match cron::Schedule::from_str(&expr) {
            Ok(s) => s,
            Err(e) => {
                LOG.error(
                    "Некорректное cron-выражение — задача не будет запускаться",
                    &[
                        f("задача", task.id.clone()),
                        f("название", task.title.clone()),
                        f("выражение", task.cron_expression.clone()),
                        f("ошибка", e),
                    ],
                );
                self.registry.set_next_run(&task.id, &task.title, task.mode.as_str(), None);
                return None;
            }
        };

        let next = schedule.upcoming(Local).next();
        match next {
            Some(when) => {
                // Задача, у которой вышел срок окончания, больше не планируется.
                if let Some(end) = task.end.0 {
                    if when > end {
                        LOG.info(
                            "Задача завершена по дате окончания",
                            &[
                                f("задача", task.id.clone()),
                                f("название", task.title.clone()),
                                f("окончание", end.format("%d.%m.%Y %H:%M")),
                            ],
                        );
                        self.registry.set_next_run(&task.id, &task.title, task.mode.as_str(), None);
                        return None;
                    }
                }
                // Фиксируем план в реестре статусов — оператор видит, когда следующий запуск.
                self.registry.set_next_run(&task.id, &task.title, task.mode.as_str(), Some(CsTime(Some(when))));
                Some(when)
            }
            None => {
                self.registry.set_next_run(&task.id, &task.title, task.mode.as_str(), None);
                None
            }
        }
    }

    /// Поток-планировщик: спит до ближайшего срабатывания, ставит задачи в очередь
    /// и сразу планирует следующий запуск (не дожидаясь завершения зонда).
    fn scheduler_loop(self: Arc<Self>) {
        loop {
            let mut tasks = self.tasks.lock().unwrap();

            // Ближайший срок среди всех задач.
            let now = Local::now();
            let mut due: Vec<String> = Vec::new();
            let mut nearest: Option<DateTime<Local>> = None;
            for (id, entry) in tasks.iter() {
                match entry.next {
                    Some(when) if when <= now => due.push(id.clone()),
                    Some(when) => nearest = Some(nearest.map_or(when, |n: DateTime<Local>| n.min(when))),
                    None => {}
                }
            }

            // Наступившие задачи — в очередь, следующий запуск планируем сразу.
            for id in due {
                let task = match tasks.get(&id) {
                    Some(entry) => Arc::clone(&entry.task),
                    None => continue,
                };
                self.enqueue(Arc::clone(&task));
                let next = self.compute_next(&task);
                if let Some(entry) = tasks.get_mut(&id) {
                    entry.next = next;
                    if let Some(when) = next {
                        nearest = Some(nearest.map_or(when, |n: DateTime<Local>| n.min(when)));
                    }
                }
            }

            // Спим до ближайшего срока (но не дольше секунды — чтобы реагировать на изменения).
            let sleep = nearest
                .map(|when| (when - Local::now()).to_std().unwrap_or(Duration::from_millis(1)))
                .unwrap_or(Duration::from_secs(1))
                .min(Duration::from_secs(1));
            let _ = self.wake.wait_timeout(tasks, sleep).unwrap();
        }
    }

    /// Идентификаторы задач по расписанию (для сверки сервером).
    pub fn known_task_ids(&self) -> Vec<String> {
        self.tasks.lock().unwrap().keys().cloned().collect()
    }

    /// Полные определения задач по расписанию (сервер забирает их для восстановления БД).
    pub fn all_tasks(&self) -> Vec<TaskInfo> {
        self.tasks.lock().unwrap().values().map(|e| (*e.task).clone()).collect()
    }

    /// Останавливает и удаляет ВСЕ задачи вместе с файлом реестра.
    /// Используется сторожем связи: сервер молчит дольше таймаута — проба считает себя
    /// удалённой. Возвращает число остановленных задач.
    pub fn clear_all(&self) -> usize {
        let mut tasks = self.tasks.lock().unwrap();
        let count = tasks.len();
        for id in tasks.keys() {
            self.registry.remove(id);
        }
        tasks.clear();
        let _ = std::fs::remove_file(TASKS_FILE);
        count
    }
}

/// Сохраняет реестр на диск; вызывается под замком.
fn persist(tasks: &HashMap<String, Scheduled>) {
    let all: Vec<&TaskInfo> = tasks.values().map(|e| e.task.as_ref()).collect();
    let data = match serde_json::to_vec(&all) {
        Ok(d) => d,
        Err(e) => {
            LOG.error("Не удалось сериализовать реестр задач", &[f("ошибка", e)]);
            return;
        }
    };
    let tmp = format!("{TASKS_FILE}.tmp");
    if let Err(e) = std::fs::write(&tmp, &data) {
        LOG.error("Не удалось сохранить реестр задач", &[f("файл", TASKS_FILE), f("ошибка", e)]);
        return;
    }
    if let Err(e) = std::fs::rename(&tmp, TASKS_FILE) {
        LOG.error("Не удалось заменить файл реестра задач", &[f("файл", TASKS_FILE), f("ошибка", e)]);
    }
}
